src/foam/dao/jdbc/AbstractJDBCDAO.java

```diff
@@ -0,0 +1,283 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.dao.jdbc;
+
+import foam.dao.AbstractDAO;
+import foam.core.*;
+import foam.nanos.logger.Logger;
+
+import javax.sql.DataSource;
+import java.sql.SQLException;
+import java.sql.*;
+import java.lang.Exception;
+import java.util.*;
+
+/**
+ * Abstract class for implementing JDBC-based DAOs.
+ */
+public abstract class AbstractJDBCDAO extends AbstractDAO{
+
+  /** Holds the relevant properties (column names) of the table */
+  protected List<PropertyInfo> properties_;
+
+  protected ThreadLocal<StringBuilder> threadLocalBuilder_ = new ThreadLocal<StringBuilder>(){
+
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder builder = super.get();
+      builder.setLength(0);
+      return builder;
+    }
+
+  };
+
+  protected String tableName_;
+
+  /** Holds a reference to the connection pool ( .getConnection() ) */
+  protected static DataSource dataSource_;
+
+  @Override
+  public FObject find_(X x, Object id) {
+    Connection c = null;
+    IndexedPreparedStatement stmt = null;
+    ResultSet resultSet = null;
+
+    try {
+      c = dataSource_.getConnection();
+      StringBuilder builder = threadLocalBuilder_.get()
+        .append("select * from ")
+        .append(tableName_)
+        .append(" where ")
+        .append(getPrimaryKey().createStatement())
+        .append(" = ?");
+
+      stmt = new IndexedPreparedStatement(c.prepareStatement(builder.toString()));
+      // TODO: add support for non-numbers
+      //stmt.setLong(((Number) o).longValue());
+      stmt.setObject(id);
+      resultSet = stmt.executeQuery();
+      if ( ! resultSet.next() ) {
+        // no rows
+        return null;
+        // In the doc, should be: throw new foam.dao.ObjectNotFoundException();
+      }
+
+      return createFObject(resultSet);
+    } catch (Throwable e) {
+      Logger logger = (Logger) x.get("logger");
+      logger.error(e);
+      return null;
+    } finally {
+      closeAllQuietly(resultSet, stmt);
+    }
+  }
+
+  @Override
+  public FObject remove_(X x, FObject obj) {
+    Connection c = null;
+    IndexedPreparedStatement stmt = null;
+
+    try {
+      c = dataSource_.getConnection();
+      StringBuilder builder = threadLocalBuilder_.get()
```

Comment:
Why do we need to build this string every time? Isn't it always the same so we could reuse it? Same as for other prepared statements.